core/src/main/scala/org/ensime/indexer/SearchService.scala

```diff
@@ -64,85 +67,85 @@ class SearchService(
-          try {
-            log.debug(s"indexing $jar")
-            val check = FileCheck(jar)
-            val symbols = scan(vfs.vjar(jar)) flatMap (extractSymbols(jar, _))
+
+      _ <- Future.sequence(
+        // individual DELETEs in H2 are really slow
+        stale.grouped(1000).map(delete)
+      )
+
+      // start indexing after all deletes have completed (not pretty)
+
+      bases = {
+        config.modules.flatMap {
+          case (name, m) =>
+            m.targetDirs.flatMap { d => scan(vfs.vfile(d)) } ::: m.testTargetDirs.flatMap { d => scan(vfs.vfile(d)) } :::
+              m.compileJars.map(vfs.vfile) ::: m.testJars.map(vfs.vfile)
+        }
+      }.toSet ++ config.javaLibs.map(vfs.vfile)
+
+      basesWithOutOfDateInfo <- Future.sequence(bases.map(b => db.outOfDate(b).map((b, _))))
+      persisted <- Future.sequence(
+        basesWithOutOfDateInfo.collect { // is there a simpler way to do this, rather than collect after sequence?
+          case (base, outOfDate) if outOfDate => base
+        }.map {
+          case classfile if classfile.getName.getExtension == "class" => Future[Future[Unit]] {
+            val check = FileCheck(classfile)
+            val symbols = extractSymbols(classfile, classfile)
             persist(check, symbols)
-          } catch {
-            case e: Exception =>
-              log.error(s"Failed to index $jar", e)
-          }
-        }(workerEC)
+          }.flatMap(identity)
+
+          case jar => Future[Future[Unit]] {
+            try {
+              log.debug(s"indexing $jar")
+              val check = FileCheck(jar)
+              val symbols = scan(vfs.vjar(jar)) flatMap (extractSymbols(jar, _))
+              persist(check, symbols)
+            } catch {
+              case e: Exception =>
+                log.error(s"Failed to index $jar", e)
+                Future.successful(())
+            }
+          }.flatMap(identity)
+        }
+      )
+
+      _ = {
+        // delayed commits speedup initial indexing time
+        log.debug("committing index to disk...")
+        index.commit()
+        log.debug("...done committing index")
       }
-    }
-
-    val indexed = indexing.flatMap { w => Future.sequence(w) }.map(_.size)
-    val indexedAndCommitted = indexed map { count =>
-      // delayed commits speedup initial indexing time
-      log.debug("committing index to disk...")
-      index.commit()
-      log.debug("...done committing index")
-      count
-    }
 
-    for {
-      r <- removed
-      i <- indexedAndCommitted
-    } yield (r, i)
+    } yield (stale.size, persisted.size)
   }
 
   def refreshResolver(): Unit = resolver.update()
 
-  def persist(check: FileCheck, symbols: List[FqnSymbol]): Unit = try {
+  def persist(check: FileCheck, symbols: List[FqnSymbol]): Future[Unit] = try {
     index.persist(check, symbols)
-    db.persist(check, symbols)
+    db.persist(check, symbols).map(_ => ())
   } catch {
     case e: SQLException =>
```

Comment:
I'm assuming that if the underlying slick layer throws an exception, it'll only be visible when we try to read the result of the Future. In which case, the below exception logic needs to move, or be attached as an `onFailure`